tests/UnitTests/BankTransactionUnitTest.php

```diff
@@ -30,4 +34,147 @@ public function testCanGetReferenceFromLine()
+        $bankTransaction->addLine($line);
+
+        return $line;
+    }
+
+    public function testGetLinesReturnsLinesInTheSameOrderAsTheyWereAdded()
+    {
+        $total_line_count = 300;
+
+        $bank_transaction = new BankTransaction();
+
+        $this->addTotalLineWithDescription($bank_transaction, '0');
+
+        for ($i = 1; $i < $total_line_count; $i++) {
+            $this->addDetailLineWithDescription($bank_transaction, (string)$i);
+        }
+
+        self::assertLinesInOrderBasedOnDescription($bank_transaction, $total_line_count);
+    }
+
+    public function testGetLinesReturnsEmptyArrayWhenNoLinesSet()
+    {
+        $bank_transaction = new BankTransaction();
+
+        self::assertCount(0, $bank_transaction->getLines());
+    }
+
+    public function testGetLinesReturnsLinesInSameOrderWhenOnlyDetailsAdded()
+    {
+        $bank_transaction = new BankTransaction();
+
+        $this->addDetailLineWithDescription($bank_transaction, '0');
+        $this->addDetailLineWithDescription($bank_transaction, '1');
+        $this->addDetailLineWithDescription($bank_transaction, '2');
+
+        self::assertLinesInOrderBasedOnDescription($bank_transaction, 3);
+    }
+
+    public function testGetLinesReturnsOnlyTotalLineWhenOnlyTotalLineIsAdded()
+    {
+        $bank_transaction = new BankTransaction();
+
+        $this->addTotalLineWithDescription($bank_transaction, '0');
+
+        self::assertCount(1, $bank_transaction->getLines());
+        self::assertTrue($bank_transaction->getLines()[0]->getLineType()->equals(LineType::TOTAL()));
+    }
+
+    public function testAdding1TotalLineAnd499DetailLinesToABankTransactionWorks()
+    {
+        $total_line_count = 500;
+
+        $bank_transaction = new BankTransaction();
+
+        $this->addTotalLineWithDescription($bank_transaction, '0');
+        for ($i = 1; $i < $total_line_count; $i++) {
+            $this->addDetailLineWithDescription($bank_transaction, (string)$i);
+        }
+
+        self::assertCount($total_line_count, $bank_transaction->getLines());
+    }
+
+    public function testAdding1TotalLineWith500DetailLinesToABankTransactionThrows()
+    {
+        $total_line_count = 501;
+
+        $bank_transaction = new BankTransaction();
+
+        $this->addTotalLineWithDescription($bank_transaction, '0');
+
+        $this->expectException(\InvalidArgumentException::class);
+
+        for ($i = 1; $i < $total_line_count; $i++) {
+            $this->addDetailLineWithDescription($bank_transaction, (string)$i);
+        }
+    }
+
+    public function testAddLineThrowsAndIgnoresWhenAddingASecondTotalLine()
+    {
+        $bank_transaction = new BankTransaction();
+
+        $this->addTotalLineWithDescription($bank_transaction, '0');
+
+        try {
+            $this->addTotalLineWithDescription($bank_transaction, '0');
+
+            self::fail('An exception should have been thrown when adding a second total line');
+        } catch (\InvalidArgumentException $e) {
```

Comment:
I thought you could just call `$this->expectException()` between the first and the second call to `addTotalLineWithDescription()`, so you don't have to do the catching yourself.

Reply:
I was going to do that, but then I noticed a flaw in my implementation.

My implementation would still add the line, and then throw because there are 2 TOTAL lines set on the bank transaction. Obviously, even when throwing, we shouldn't actually set the second TOTAL transaction.

Thus I'm catching the exception here and asserting that there is still only 1 transaction.